clippy_lints/src/transmute/missing_transmute_annotations.rs

```diff
@@ -0,0 +1,53 @@
+use rustc_errors::Applicability;
+use rustc_hir::{GenericArg, HirId, Node, Path, TyKind};
+use rustc_lint::LateContext;
+use rustc_middle::lint::in_external_macro;
+use rustc_middle::ty::Ty;
+
+use clippy_utils::diagnostics::span_lint_and_sugg;
+
+use crate::transmute::MISSING_TRANSMUTE_ANNOTATIONS;
+
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    path: &Path<'tcx>,
+    from_ty: Ty<'tcx>,
+    to_ty: Ty<'tcx>,
+    expr_hir_id: HirId,
+) -> bool {
+    let last = path.segments.last().unwrap();
+    if in_external_macro(cx.tcx.sess, last.ident.span) {
+        // If it comes from a non-local macro, we ignore it.
+        return false;
+    }
+    let args = last.args;
+    let missing_generic = match args {
+        Some(args) if !args.args.is_empty() => args.args.iter().any(|arg| match arg {
+            GenericArg::Infer(_) => true,
+            GenericArg::Type(ty) => matches!(ty.kind, TyKind::Infer),
+            _ => false,
+        }),
+        _ => true,
+    };
+    if !missing_generic {
+        return false;
+    }
+    // If it's being set as a local variable value...
```

Comment:
This doesn't currently check that only the output type parameter was `_`. So for example this is not linted
```rs
let _: i64 = transmute([x, y]);
```
despite the missing input type annotation.

Reply:
It should presumably also look past ancestor blocks, so that `let _: T = unsafe { transmute::<_, _>() };` is linted. The parent expression of transmute is the unsafe block, not the let binding.

Reply:
Though I suppose allowing blocks might be getting close to the closure case pointed out by Centri, or other single-expression functions.
It does feel a bit arbitrary to allow specific cases like these but forbid closures/blocks though.